### ISSUE TITLE
Custom Cloud Environment attempting authentication to .com endpoint

### ISSUE THREAD

**Issue by Smgilliam25** [hashicorp/terraform-provider-azurerm#22495]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.2.4

### AzureRM Provider Version

3.51.0

### Affected Resource(s)/Data Source(s)

azurerm

### Terraform Configuration Files

```hcl
provider "azurerm"
metadata_host = "management.azure.microsoft.x" (Custom Cloud)
version 3.51.0
```

### Debug Output/Panic Output

```shell
The Resource Principal named "Management.azure.com" was not found in the tenant named X. This can happen if the application has not been installed by the admin or you may have sent your authentication request to the wrong tenant.
```

### Expected Behaviour

The provider should be using .cloud domain endpoints opposed to the .com endpoints, this is working with provider 2.85.0 however we require newer releases to support Gen2 and Virtualization Based Security Features. 

### Actual Behaviour

The Service Principal login fails because the resource principal named "management.azure.com" does not exist. Developer error - the app requested access to a resource (application) that isn't installed in your tenant. 

### Steps to Reproduce

Terraform plan

### Important Factoids

Custom Secure Cloud Environment 

### References

https://github.com/hashicorp/terraform-provider-azurerm/issues/15743
8

**Comment by magodo** [comment]
@Smgilliam25 Sorry for hitting this issue. Would you please provide the debug log so that I can better understand what's going on?

By saying:

> The provider should be using .cloud domain endpoints opposed to the .com endpoints

Can you elaborate more about your expectation here? Isn't `management.azure.com` SP something users specified?

**Comment by Smgilliam25** [comment]
> @Smgilliam25 Sorry for hitting this issue. Would you please provide the debug log so that I can better understand what's going on?
> 
> By saying:
> 
> > The provider should be using .cloud domain endpoints opposed to the .com endpoints
> 
> Can you elaborate more about your expectation here? Isn't `management.azure.com` SP something users specified?

As this is a highly secure environment I have scrubbed a few of the details...

My Debug Efforts on a working provider 2.85.0 vs a Non working 3.51.0 are as follows:

Version 3.51.0: (Non Working)
My Debug Shows a Post to https://login.microsoftonline.microsoft.xxdomain/id/oauth2/v2.0/token/timestamp and pulls the following error: 
Error retrieving providers: Listing Resource Providers: autorest/Client: Preparing request failed. Error: client CredentialsToken: received HTTP status 400 with response: { " error: invalid_resource", "error_Description AADSTS500011: The resource Principal named https://management.azure.com/ was not found in the tenant named "xTenant". This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You might have sent your authentication request to the wrong tenant. 

This is a non reachable endpoint and should be routing to management.azure.microsoft.xdomain based off the "metadata_host" in my providers file...It seems that everything is working up until this point as it pulls the correct login endpoint.

2.85.0 (Working Version) 
My debug runs through testing if Service Principal/Client Secret is applicable for authentication, getting Oauth config for endpoint login.microsoftonline.microsoft.xdomain w/ tenant x, then runs a "GoAzureHelpers Request" Post to login and pulls an oauth token. 



-

**Comment by Smgilliam25** [comment]
FSA, I imported ARM 2.88.0 and it appears that I am able to authenticate and use the commands I need (Secure Boot). It appears that the issues may be with the switch from AD auth to MSAL, however even with disabling MSAL in my providers configuration, ARM is still attempting to reach the .com endpoint for ARM vs the correct inboundary resource pulled from the metadata resource.

**Comment by magodo** [comment]
@manicminer Would you like to take a look at this one?

**Comment by manicminer** [comment]
@Smgilliam25 Thanks for reporting this. As you might imagine, it's challenging to test or get feedback for the provider running in custom clouds, so your input is greatly appreciated.

Before I'm able to look into this more closely, can you advise which authentication method you're using? To my knowledge, our authentication library does not call that endpoint, so I wondered if perhaps you're using Azure CLI for authentication?

**Comment by manicminer** [comment]
@Smgilliam25 I've taken a look at the authentication workflow and I haven't yet spotted the likely cause of your issue. Can you confirm which authentication method you're using (az-cli, native service principal auth, or OIDC)? This will greatly help narrow down the root cause. Thanks!

**Comment by UnexpectedAdmin** [comment]
@manicminer I've run into the same issue as described above.

Authentication is being done via the az-cli.   `az account show` shows proper authentication to the private cloud.
I've set the following environment variables:
```
ARM_CLOUD_METADATA_URL
ARM_METADATA_HOSTNAME
```

Debugging Notes:
- via tcpdump, I'm able to see https packets going to/from the proper private cloud endpoint (ex: management.azure.local)
- can curl the metadata URL and made sure it supports schema version >= 2022-09-01 and has required fields (loginEndpoint, resourceManager, microsoftGraphResourceID)
- ~/.azure/cloudEndpoints.json looks correct
- I am authenticating via a user account that has proper admin privileges (can create/modify Azure resources)

Terraform Version
1.5.7
AzureRM Provider Version
3.75.0
AZ-CLI Version
2.38.1

**Comment by manicminer** [comment]
@UnexpectedAdmin Thanks for the feedback. Could you post a debug log for the interaction covering the 400 error to the token endpoint? This will help greatly in determining the root cause of the issue, as we don't have ready access to a private cloud environment with which to reproduce. Thanks!

**Comment by Smgilliam25** [comment]
> @Smgilliam25 I've taken a look at the authentication workflow and I haven't yet spotted the likely cause of your issue. Can you confirm which authentication method you're using (az-cli, native service principal auth, or OIDC)? This will greatly help narrow down the root cause. Thanks!

I am attempting authentication via Service Principal. Sorry for the delay here.

**Comment by manicminer** [comment]
@Smgilliam25 Thanks for reporting back. I've dug a little further and it's possible I _may_ have found the source of this issue. Would you be able to locally build and run the provider from the [`bugfix/custom-metadata-resource-identifiers`](https://github.com/hashicorp/terraform-provider-azurerm/compare/bugfix/custom-metadata-resource-identifiers?expand=1) branch to test my hypothesis?

**Comment by UnexpectedAdmin** [comment]
@manicminer Awesome! We will give this a shot, but probably not until after Thanksgiving.

**Comment by rpfinneran** [comment]
@manicminer - I am helping @Smgilliam25 out on this. We did the build from the identified branch and have tested against commercial endpoints successfully. We're working on transferring to the secure environment to test, it may take some time though. Will update once it is completed.

**Comment by manicminer** [comment]
@rpfinneran That's great, thanks for the update and appreciate your taking the time to test it 👍

**Comment by UnexpectedAdmin** [comment]
@manicminer The bugfix branch appears to resolve our issue. We are able to deploy resources with it in the private cloud.
Thank you!

**Comment by jasbryan** [comment]
Do we have any ETA on when this bugfix can get into the main release for the azurerm provider?

**Comment by toefraz** [comment]
@manicminer thanks for getting this merged upstream. Do you think the SDK update will make it here into 3.84.0?

**Comment by randomnote1** [comment]
Looks like 3.84.0 works!

**Comment by tombuildsstuff** [comment]
Closing since per @randomnote1's comment this has been resolved when `hashicorp/go-azure-sdk` was updated.

**Comment by manicminer** [comment]
Just a note to custom cloud users, we fixed a minor bug in last week's release v3.101.0 and it's our intention and belief that all custom cloud environments should now be fully supported.

It's worth mentioning that since v3.101.0, the only provider that needs to be set is `metadata_host`. It should no longer be necessary to set `environment` and this will in fact be ignored when using a custom metadata host.

Related issue/PR: https://github.com/hashicorp/terraform-provider-azurerm/issues/25675 / https://github.com/hashicorp/terraform-provider-azurerm/pull/25717
